src/combinatorics/derangements/basic.lean

```diff
@@ -24,105 +25,134 @@ We also define:
 
-/-- The permutation `f` has at most one fixed point, `a`. -/
-def only_possible_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x → x = a
+variables {α β : Type*}
 
-/-- The permutation `f` has exactly one fixed point, `a`. -/
-def exactly_one_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x ↔ x = a
-
-end definitions
-
-section simple_lemmas
+lemma mem_derangements_iff_fixed_points_eq_empty {f : perm α} :
+  f ∈ derangements α ↔ fixed_points f = ∅ :=
+set.eq_empty_iff_forall_not_mem.symm
 
 /-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
-protected def congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
-subtype_equiv (perm_congr e) $ λ f, e.forall_congr $ λ x, by simp
-
-end simple_lemmas
-
-section fixed_points
-
-variables {α : Type*} [decidable_eq α]
-
-lemma eofp_iff_opfp_and_eq (a : α) (f : perm α) :
-  exactly_one_fixed_point f a ↔ only_possible_fixed_point f a ∧ f a = a :=
-⟨λ h, ⟨λ x, (h x).mp, (h a).mpr rfl⟩, λ h x, ⟨h.1 x, forall_eq.mpr h.2 x⟩⟩
-
-lemma mem_derangements_iff_opfp_and_ne (a : α) (f : perm α) :
-  f ∈ derangements α ↔ only_possible_fixed_point f a ∧ f a ≠ a :=
-begin
-  refine ⟨λ h_derangement, ⟨λ x x_fixed, (h_derangement x x_fixed).elim, h_derangement a⟩, _⟩,
-  rintros ⟨h_opfp, fa_ne_a⟩ x x_fixed,
-  specialize h_opfp x x_fixed,
-  rw h_opfp at x_fixed,
-  exact fa_ne_a x_fixed
-end
+def equiv.derangements_congr (e : α ≃ β) : (derangements α ≃ derangements β) :=
+e.perm_congr.subtype_equiv $ λ f, e.forall_congr $ λ x, by simp
 
-/-- The set of permutations fixing `a` is the same as the set of permutations on `{a}ᶜ`. -/
-def discard_fixed_pt (a : α) : {f : perm α | f a = a} ≃ perm ({a}ᶜ : set α) :=
-begin
-  refine (subtype_equiv_right _).trans (equiv.set.compl (equiv.refl _)),
-  simp
-end
+namespace derangements
 
-/-- The set of permutations with `a` the only fixed point is equivalent to the set of derangements
-    on `{a}ᶜ`. -/
-def eofp_equiv_derangements_except_for (a : α) :
-  {f : perm α // exactly_one_fixed_point f a} ≃ derangements ({a}ᶜ : set α) :=
-begin
-  transitivity {f : {f : perm α // f a = a} // only_possible_fixed_point f.val a},
-  { refine (subtype_equiv_right _).trans (subtype_subtype_equiv_subtype_exists _ _).symm,
-    intro f,
-    simp_rw [exists_prop, eofp_iff_opfp_and_eq, and_comm] },
-  { refine subtype_equiv (discard_fixed_pt a) _,
-    rintro ⟨f, _⟩,
-    simp [discard_fixed_pt, equiv.set.compl, derangements,
-          only_possible_fixed_point, not_imp_not] }
-end
+/-- Derangements on a subtype are equivalent to permutations on the original type whose set of fixed
+points is the rest. -/
+protected def subtype_equiv  (p : α → Prop) [decidable_pred p] :
+  derangements (subtype p) ≃ {f : perm α // ∀ a, ¬p a ↔ a ∈ fixed_points f} :=
+calc
+  derangements (subtype p)
+      ≃ {f : {f : perm α // ∀ a, ¬p a → a ∈ fixed_points f} // ∀ a, a ∈ fixed_points f → ¬p a}
+      : begin
```

Comment:
is there any difference between `by {...}` and `begin ... end`? just stylistic?

Reply:
They are strictly the same thing!